File: app/archive-swift/RetroRaceCore/Library.swift
import CommonCrypto
import Foundation

// MARK: - Local library: detected games

/// A locally-owned game file, identified by hash — the ROM never leaves the
/// machine (roadmap: local library + hash import).
public struct GameEntry: Identifiable, Hashable {
    public let id: String          // sha256
    public let name: String        // filename without extension
    public let path: String
    public let md5: String
    public let sha256: String
    public let size: Int
    public let extensions: String  // e.g. "nes"

    /// Best-effort display name: strip the extension and common archive noise.
    public var displayName: String {
        name
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "_", with: " ")
    }

    public init(id: String, name: String, path: String, md5: String,
                sha256: String, size: Int, extensions: String) {
        self.id = id
        self.name = name
        self.path = path
        self.md5 = md5
        self.sha256 = sha256
        self.size = size
        self.extensions = extensions
    }
}

/// Scans a folder for supported ROMs and hashes each file (local-only).
public struct LocalLibrary {
    /// Extensions mapped to the core dylib path, relative to the cores dir.
    public static let supportedExtensions = [
        "nes": "libretro-fceumm/fceumm_libretro.dylib",
        "fds": "libretro-fceumm/fceumm_libretro.dylib",
        "snes": "snes9x_libretro.dylib",
        "sfc": "snes9x_libretro.dylib",
        "smc": "snes9x_libretro.dylib",
        "gb": "gambatte_libretro.dylib",
        "gbc": "gambatte_libretro.dylib",
        "gen": "genesis_plus_gx_libretro.dylib",
        "md": "genesis_plus_gx_libretro.dylib",
    ]

    public static func scan(_ directory: URL) -> [GameEntry] {
        let fm = FileManager.default
        guard let files = try? fm.contentsOfDirectory(at: directory,
                                                      includingPropertiesForKeys: [.fileSizeKey],
                                                      options: [.skipsHiddenFiles]) else {
            return []
        }

        return files
            .filter { !$0.hasDirectoryPath }
            .compactMap { url -> GameEntry? in
                let ext = url.pathExtension.lowercased()
                guard supportedExtensions[ext] != nil else { return nil }
                guard let data = try? Data(contentsOf: url) else { return nil }

                let md5 = md5Hex(data)
                let sha256 = sha256HexData(data)
                let size = (try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? data.count
                return GameEntry(id: sha256,
                                 name: url.deletingPathExtension().lastPathComponent,
                                 path: url.path,
                                 md5: md5,
                                 sha256: sha256,
                                 size: size,
                                 extensions: ext)
            }
            .sorted { $0.displayName.localizedCaseInsensitiveCompare($1.displayName) == .orderedAscending }
    }
}

func md5Hex(_ data: Data) -> String {
    var digest = [UInt8](repeating: 0, count: Int(CC_MD5_DIGEST_LENGTH))
    data.withUnsafeBytes { buf in
        _ = CC_MD5(buf.baseAddress, CC_LONG(data.count), &digest)
    }
    return digest.map { String(format: "%02x", $0) }.joined()
}

func sha256HexData(_ data: Data) -> String {
    var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    data.withUnsafeBytes { buf in
        _ = CC_SHA256(buf.baseAddress, CC_LONG(data.count), &digest)
    }
    return digest.map { String(format: "%02x", $0) }.joined()
}

File: app/internal/library/library.go
package library

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// Console is one supported system with its games.
type Console struct {
	ID    string // "nes", "snes", ...
	Name  string // "NES / Famicom", ...
	Core  string // relative dylib path under the cores dir
	Exts  []string
	Games []Game
	Emoji string
}

// Game is a locally-owned ROM identified by hashes.
type Game struct {
	Name    string // display name (no extension)
	Path    string
	SHA256  string
	Size    int64
	Ext     string
	Console string // console id
}

// Library is the console registry for a run, populated from a ROM directory.
type Library struct {
	Consoles []Console
}

// New returns a Library with the supported console registry. Core paths are
// relative to the cores directory.
func New() *Library {
	return &Library{Consoles: []Console{
		{ID: "nes", Name: "NES / Famicom", Core: "libretro-fceumm/fceumm_libretro.dylib", Exts: []string{"nes", "fds"}, Emoji: "🕹️"},
		{ID: "snes", Name: "Super Nintendo / SNES", Core: "snes9x_libretro.dylib", Exts: []string{"sfc", "smc", "snes"}, Emoji: "🎮"},
		{ID: "gb", Name: "Game Boy / Color", Core: "gambatte_libretro.dylib", Exts: []string{"gb", "gbc"}, Emoji: "📟"},
		{ID: "genesis", Name: "Genesis / Mega Drive", Core: "genesis_plus_gx_libretro.dylib", Exts: []string{"gen", "md"}, Emoji: "⚡"},
	}}
}

// Scan scans a ROM directory and returns the consoles that have matching
// games, sorted games-first (alphabetical). It mutates only l, never any
// package-level state.
func (l *Library) Scan(romsDir string) []Console {
	entries, err := os.ReadDir(romsDir)
	if err != nil {
		return nil
	}

	extToConsole := map[string]*Console{}
	for i := range l.Consoles {
		for _, e := range l.Consoles[i].Exts {
			extToConsole[e] = &l.Consoles[i]
		}
	}

	// Reset games.
	for i := range l.Consoles {
		l.Consoles[i].Games = nil
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		ext := strings.ToLower(strings.TrimPrefix(filepath.Ext(entry.Name()), "."))
		console, ok := extToConsole[ext]
		if !ok {
			continue
		}
		path := filepath.Join(romsDir, entry.Name())
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		sha := sha256.Sum256(data)
		name := strings.TrimSuffix(entry.Name(), filepath.Ext(entry.Name()))
		console.Games = append(console.Games, Game{
			Name:    strings.ReplaceAll(name, "_", " "),
			Path:    path,
			SHA256:  hex.EncodeToString(sha[:]),
			Size:    int64(len(data)),
			Ext:     ext,
			Console: console.ID,
		})
	}

	// Sort consoles with games first, games alphabetically.
	nonEmpty := make([]Console, 0, len(l.Consoles))
	for i := range l.Consoles {
		if len(l.Consoles[i].Games) > 0 {
			sort.Slice(l.Consoles[i].Games, func(a, b int) bool {
				return l.Consoles[i].Games[a].Name < l.Consoles[i].Games[b].Name
			})
			nonEmpty = append(nonEmpty, l.Consoles[i])
		}
	}
	return nonEmpty
}
